### ISSUE TITLE
SC2046: Don't react to `set --` command

### ISSUE THREAD

**Issue by AlexanderAmelkin** [koalaman/shellcheck#2335]
#### For bugs
- Rule Id: SC2046
- My shellcheck version: 0.3.7
- [X] The rule's wiki page does not already cover this
- [X] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

function get_gpio()
{
  local chip
  local pin
  # We do want word splitting here, in fact the `set --` command
  # is specifically designed for word splitting
  set -- $(gpiofind ${INTR_RST_GPIO_PIN})
  chip=$1
  pin=$2
# some more code
}
```
#### Here's what shellcheck currently says:

```
In myscript line 9:
  set -- $(gpiofind ${INTR_RST_GPIO_PIN})
         ^-- SC2046: Quote this to prevent word splitting.
```

#### Here's what I wanted or expected to see:

No warnings for dedicated word-splitting `set --` command.

**Comment by AlexanderAmelkin** [comment]
This is probably a partial duplicate of #1284, but in this one I would like to emphasize that `set --` was designed for word splitting.
I would rather complain if it was used with quotes because that would probably produce a wrong result.

**Comment by kurahaupo** [comment]
My most common use is simply:

    set -- --foo "$@"

where I certainly would _not_ want word-splitting.

This is one of those cases where it's entirely appropriate to annotate the code with

    # shellcheck disable=SC2046
    set -- $( gpiofind "$INTR_RST_GPIO_PIN" )

(Did you intend to omit both pairs of quotes, or only the outer pair?)

**Comment by kurahaupo** [comment]
@AlexanderAmelkin perhaps a better way to implement `get_gpio` would be get `read` to do the wordsplitting:
```
#!/bin/bash

get_gpio()
{
  local chip pin
  IFS=' ' read chip pin _ < <(
    gpiofind "$INTR_RST_GPIO_PIN"
  )
# some more code
}
```

**Comment by mattmc3** [comment]
> This is one of those cases where it's entirely appropriate to annotate the code with
> 
> ```
> # shellcheck disable=SC2046
> set -- $( gpiofind "$INTR_RST_GPIO_PIN" )
> ```

I disagree, and would like to see this scenario split out from SC2046 into its own separate check. That way, a script with lots of `set -- $(...)` calls won't need to be littered with `#shellcheck disable=SC2046` annotations. `set -- $(...)` happens all the time in POSIX sh scripts since it's the only way to have arrays. If it was a separate check, you could simply ignore it at the top for scripts that fit that pattern, and be on your merry way with SC2046 still correctly checking for other word splitting scenarios.

**Comment by kurahaupo** [comment]
@mattmc3 just because a method is "common" doesn't mean it's the "only" method, and it certainly doesn't mean that it's "safe".

As an alternative, I'm pretty sure that POSIX `read` does splitting as I've already described, and without the risk of inadvertent globbing.

I think that treating `set -- $(cmd)` as a special case would be going in the wrong direction. A more general - and safer - solution would be to add a mechanism to nominate `cpiofind` as being a command that produces split-and-glob-safe output, perhaps with some additional notation as to what value of `IFS` is expected. (e.g. `getent passwd` would be safe when `IFS=:`.)

That way you'd be forced to provide a comprehensive list of commands whose output you'd audited, so weird things can't slip through unnoticed, but you'd only have to write the list once, rather than annotate every `set` statement.

A similar method might be applied to variables, so for example unquoted `$PATH` would be deemed safe when `IFS=:`.

**Comment by kurahaupo** [comment]
... and anyone who adds `ls` to their list of "safe" commands would get a formal reprimand on their employment record.

**Comment by krupan** [comment]
@kurahaupo you state that you are, "pretty sure that POSIX read does splitting," but can you provide an example of that?  If not, then I'm here to also ask that shellcheck not complain about `set --` in POSIX `#!/bin/sh` scripts.

**Comment by wileyhy** [comment]
POSIX-2024:

  "If IFS is unset, or is set to a non-empty value, then a modified version
of the field splitting algorithm specified in 2.6.5 Field Splitting shall
be applied with the modifications as follows...."

**Comment by kurahaupo** [comment]
@krupan The behaviour of `read` is specified by [POSIX §3](https://pubs.opengroup.org/onlinepubs/9799919799/utilities/read.html); the part describing splitting is this:

> If IFS is unset, or is set to any non-empty value, then a modified version of the field splitting algorithm specified in [2.6.5 Field Splitting](https://pubs.opengroup.org/onlinepubs/9799919799/utilities/V3_chap02.html#tag_19_06_05) shall be applied, with the modifications as follows:
> *  The input to the algorithm shall be the logical line (minus terminating delimiter) that was read from standard input, and shall be considered as a single initial field, all of which resulted from expansions, with any escaped byte and the preceding <backslash> escape character treated as if they were the result of a quoted expansion, and all other bytes treated as if they were the results of unquoted expansions.
> * The loop over the contents of that initial field shall cease when either the input is empty or n output fields have been generated, where n is one less than the number of var operands passed to the read utility. Any remaining input in the original field being processed shall be returned to the read utility "unsplit"; that is, unmodified except that any leading or trailing IFS white space, as defined in [2.6.5 Field Splitting](https://pubs.opengroup.org/onlinepubs/9799919799/utilities/V3_chap02.html#tag_19_06_05), shall be removed.

In Bash we tend to write:
```bash
IFS=,: read -r var1 var2 var3 < <(
  your_command
)
```
but `<(...)` is a Bashism. For POSIX scripts one can write:
```bash
IFS=,: read -r var1 var2 var3 <<'E'
$( your_command )
E
```
For shells that support `set -o lastpipe`, we can also write
```bash
set -o lastpipe
your _command |
IFS=,: read -r var1 var2 var3
```

One last point: if one intends to continue using `set -- $( some_cmd )`, preface that with `set -f` to prevent globbing.
